### ISSUE TITLE
Left join on non-existing data returns nothing

### ISSUE THREAD

**Issue by frankvollebregt** [simolus3/moor#1072]
**Describe the bug**

In my (Dart) query, I use joins to join some tables together. It seems that when a table is left joined in which no data exists (that corresponds with the joining criterion), the query returns no row at all. Example:

There are 3 tables: `Car`, `Engine` and `FuelTank`. In this example, `Car` has a `Car.id`. For each car, there is always an `Engine`, which references the car by its `id`. There may also be a `FuelTank`, which also contains a FK reference to the car by its `id`. Sometimes, however, no `FuelTank` is present (say the car is electric).

To the issue then: Say I execute the following Dart query:
```
final query = select(car).join([
    leftOuterJoin(engine, engine.carId.equalsExp(car.id)),
    leftOuterJoin(fuelTank, fuelTank.carId.equalsExp(car.id)),
  ]);
```
**My expectation** would be that this returns at least all `Car`s, with their `Engine`s, and if present also includes the `FuelTank`s, in correspondance with 'regular' SQL behavior.

(note: I use `readTable` to extract the different tables from the query result)

**What actually happens**, however, is that no rows at all are returned from the query. If I exclude the join with `FuelTank`, all rows (`Car` + `Engine`) return as expected, but when I include it no rows (`[]`) are returned.

Is this an intentional/known characteristic of Moor, or is there moor (heh) going on?

Or should it be working, and is there likely a mistake elsewhere in my code?

Thank you!

**Comment by simolus3** [comment]
Thank you for that excellent issue description :+1: 

Your code looks correct to me, I think it should return all cars. You can try setting `logStatements: true` on the `VmDatabase` or `FlutterQueryExecutor` to see what kind of SQL is generated here. But that really should include all rows, I'll try to reproduce your example and see where things go wrong.

**Comment by simolus3** [comment]
Sorry, I couldn't reproduce this with the latest moor version. I'm creating two cars (one with a fuel tank and one without):

```dart
class Cars extends Table {
  IntColumn get id => integer().autoIncrement()();
}

class Engines extends Table {
  IntColumn get carId => integer()();

  @override
  Set<Column> get primaryKey => {carId};
}

class FuelTanks extends Table {
  IntColumn get carId => integer()();

  @override
  Set<Column> get primaryKey => {carId};
}

@UseMoor(tables: [Cars, Engines, FuelTanks])
class Db extends _$Db {
  Db(QueryExecutor executor) : super(executor);

  @override
  int get schemaVersion => 1;

  @override
  MigrationStrategy get migration {
    return MigrationStrategy(
      beforeOpen: (details) async {
        // Car with fuel tank
        final first = await into(cars).insert(CarsCompanion.insert());
        await into(engines)
            .insert(EnginesCompanion.insert(carId: Value(first)));
        await into(fuelTanks)
            .insert(FuelTanksCompanion.insert(carId: Value(first)));

        // Car without fuel tank
        final second = await into(cars).insert(CarsCompanion.insert());
        await into(engines)
            .insert(EnginesCompanion.insert(carId: Value(second)));
      },
    );
  }

  Future<List<TypedResult>> get fullCars {
    final query = select(cars).join([
      leftOuterJoin(engines, engines.carId.equalsExp(cars.id)),
      leftOuterJoin(fuelTanks, fuelTanks.carId.equalsExp(cars.id)),
    ]);

    return query.get();
  }
}
```

For me, opening that database and printing the length of `fullCars` returns two, indicating that the query is working as expected. Are you applying another where condition after the join? If your where clause references a `NULL` table then I think that row wouldn't be included in the results.

**Comment by frankvollebregt** [comment]
Alright, so after struggling with it for a bit, everything is now working as it should be. It turned out I had made a mistake in the table declaration. While adding the foreign key constraint, I left out a closing bracket `)`, which caused the creation query to fail, and as a result the table did not exist when I tried to join it. Not sure how I got as far as I did, but now it's working as expected: If no entry exists, `null` is returned.

That is:
```
class FuelTanks extends Table {
  IntColumn get carId => integer().customConstraint("REFERENCES cars(id")();

  @override
  Set<Column> get primaryKey => {carId};
}
```
In which I forgot to add the closing bracket after `cars(id` within the custom constraint.

Thanks for your help!
